crates/sui-core/src/traffic_controller.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2021, Facebook, Inc. and its affiliates
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::net::SocketAddr;
+use std::{collections::HashMap, sync::Arc};
+
+use chrono::{DateTime, Utc};
+use mysten_metrics::spawn_monitored_task;
+use parking_lot::RwLock;
+use sui_types::traffic_control::{
+    Policy, PolicyConfig, PolicyResponse, TrafficControlPolicy, TrafficTally,
+};
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::error::TrySendError;
+use tracing::warn;
+
+type BlocklistT = Arc<RwLock<HashMap<SocketAddr, DateTime<Utc>>>>;
+
+#[derive(Clone)]
+struct Blocklists {
+    connection_ips: BlocklistT,
+    proxy_ips: BlocklistT,
+}
+
+#[derive(Clone)]
+pub struct TrafficController {
+    tally_channel: mpsc::Sender<TrafficTally>,
+    blocklists: Blocklists,
+    //metrics: TrafficControllerMetrics, // TODO
+}
+
+impl TrafficController {
+    pub async fn spawn(policy_config: PolicyConfig) -> Self {
+        let (tx, rx) = mpsc::channel(policy_config.channel_capacity);
+        let ret = Self {
+            tally_channel: tx,
+            blocklists: Blocklists {
+                connection_ips: Arc::new(RwLock::new(HashMap::new())),
+                proxy_ips: Arc::new(RwLock::new(HashMap::new())),
+            },
+        };
+        let blocklists = ret.blocklists.clone();
+        spawn_monitored_task!(run_tally_loop(rx, policy_config, blocklists));
+        ret
+    }
+
+    pub fn tally(&self, tally: TrafficTally) {
+        // Use try_send rather than send mainly to avoid creating backpressure
+        // on the caller if the channel is full, which may slow down the critical
+        // path. Dropping the tally on the floor should be ok, as in this case
+        // we are effectively sampling traffic, which we would need to do anyway
+        // if we are overloaded
+        match self.tally_channel.try_send(tally) {
+            Err(TrySendError::Full(_)) => {
+                warn!("TrafficController tally channel full, dropping tally");
+                // TODO: metric
+            }
+            Err(TrySendError::Closed(_)) => {
+                panic!("TrafficController tally channel closed unexpectedly");
+            }
+            Ok(_) => {}
+        }
+    }
+
+    pub async fn check(
+        &self,
+        connection_ip: Option<SocketAddr>,
+        proxy_ip: Option<SocketAddr>,
+    ) -> bool {
+        match (connection_ip, proxy_ip) {
+            (Some(connection_ip), _) => self.check_and_clear_blocklist(connection_ip, true).await,
+            (_, Some(proxy_ip)) => self.check_and_clear_blocklist(proxy_ip, false).await,
+            _ => true,
+        }
+    }
+
+    async fn check_and_clear_blocklist(&self, ip: SocketAddr, connection_ips: bool) -> bool {
```

Comment:
can you make the ip and `enum { ConnectionIp(SocketAddr), ProxyIp(SocketAddr) }` instead of using bool params

Reply:
reminder on this one